rfcs/20240531-coo-sparse-encoding/README.md

```diff
@@ -0,0 +1,152 @@
+
+|   | CSR Buffers    | Contents |
+|:--|:---------------|:---------|
+|0  | `data`	  | Non-zero matrix values sorted in row-major order |
+|1  | `indices`   | Column indices corresponding to each non-zero value | 
+|2  | `pointers`  | Compressed row indices - each $n^{th}$ value denotes the number of non-zero elements between $n^{th}$ and $(n+1)^{th}$ row |
+|||
+
+Because of the compressed row, this format tends to be more efficient for large sparse matrices.
+COO, on the other hand, is simpler in implementation in that the encoding comprises of a list of thruples `(values, row_index, column_index)` corresponding to the non-zero values.
+This makes the COO less efficient than CSR but it has the advantage of a reduced conversion overhead and better interpretability. 
+For practical cases, a sorted variant of COO is used wherein in the data is encoded as a set of sorted arrays containing the values, row indices and column indices respectively:
+
+|   | Sorted COO Buffers    | Contents |
+|:--|:----------------------|:---------|
+|0  | `data`	    | Non-zero values in the matrix sorted in row-major order |
+|1  | `dim0_indices`  | Dimension 0 (Row) indices for the non-zero elements | 
+|2  | `dim1_indices`  | Dimension 1 (Column) indices for the non-zero elements |
+|||
+
+## Proposal
+The proposal is to extend the oneDNN sparse memory descriptor to include support for the COO encoding. Since the sparse memory API is an experimental feature in oneDNN, the extension will also be enabled by specifying `ONEDNN_EXPERIMENTAL_SPARSE=ON`.
+
+The addition of a new encoding will update the list for the sparse format kind: 
+
+```cpp
+/// Sparse encodings
+typedef enum {
+    /// Undefined sparse encoding kind, used for empty memory descriptors.
+    dnnl_sparse_encoding_undef = 0,
+    /// Compressed Sparse Row (CSR) encoding.
+    dnnl_csr,
+    /// An encoding that is used for an opaque storage schema for
+    /// tensors with unstructured sparsity. A memory descriptor with the
+    /// packed encoding cannot be used to create a memory object. It can
+    /// only be used to create a primitive descriptor to query the
+    /// actual memory descriptor (similar to the format tag `any`).
+    dnnl_packed,
+	/// Coordinate Sparse Encoding (COO)
+	dnnl_coo,
+} dnnl_sparse_encoding_t;
+```
+
+The memory object implementation in oneDNN supports underlying memory buffers that allows it to handle non-trivial data representations like sparse encodings.
+Based on this implementation, COO encoding will have a separate API for creating a memory descriptor which can specify its buffers. This can be done in three ways:
+
+### Option 1 - Sorted COO (Recommended)
+The memory descriptors will comprise of three buffers as defined for a sorted COO with one buffer each for data (index [0]), indices for the first dimension (index [1]) and indices for the second dimension (index [2]). The buffers for the data indices will share the datatype during declaration. 
```

Comment:
It seems that API examples should be extended to `get_size` and `query_data_type` functions as for different options, it looks to me, those might look or act differently... As well as memory creation functions. Clear advantages of the recommended option should be underlined.

Reply:
Thanks for the suggestion! I have added usage examples for the `query_size()`, `query_datatype()` and memory creation functions. The advantages of the different options have also been stated.